[UI-side compositing] Resizing window doesn't cause scrollbars to show/change size

nmoucht opened the pull request:
#### 9389cc84fa302b8971ba56ab28396eb79eb33202
<pre>
[UI-side compositing] Resizing window doesn&apos;t cause scrollbars to show/change size
<a href="https://bugs.webkit.org/show_bug.cgi?id=253739">https://bugs.webkit.org/show_bug.cgi?id=253739</a>
rdar://106385267

Reviewed by Simon Fraser.

Hook up ScrollerMac to UI-process resize events so that scrollbars
dynamically resize when resizing window. We should also consider subclassing
ScrollerMac under ScrollbarsControllerMac so we dont duplicate this
NScrollerImp interaction code.

* Source/WebCore/page/scrolling/ScrollingTree.cpp:
(WebCore::ScrollingTree::viewWillStartLiveResize):
(WebCore::ScrollingTree::viewWillEndLiveResize):
(WebCore::ScrollingTree::contentsSizeChanged):
* Source/WebCore/page/scrolling/ScrollingTree.h:
* Source/WebCore/page/scrolling/ScrollingTreeFrameScrollingNode.h:
* Source/WebCore/page/scrolling/ScrollingTreeScrollingNodeDelegate.h:
(WebCore::ScrollingTreeScrollingNodeDelegate::viewWillStartLiveResize):
(WebCore::ScrollingTreeScrollingNodeDelegate::viewWillEndLiveResize):
(WebCore::ScrollingTreeScrollingNodeDelegate::contentsSizeChanged):
* Source/WebCore/page/scrolling/mac/ScrollerPairMac.h:
* Source/WebCore/page/scrolling/mac/ScrollerPairMac.mm:
(WebCore::ScrollerPairMac::viewWillStartLiveResize):
(WebCore::ScrollerPairMac::viewWillEndLiveResize):
(WebCore::ScrollerPairMac::contentsSizeChanged):
* Source/WebCore/page/scrolling/mac/ScrollingTreeScrollingNodeDelegateMac.h:
* Source/WebCore/page/scrolling/mac/ScrollingTreeScrollingNodeDelegateMac.mm:
(WebCore::ScrollingTreeScrollingNodeDelegateMac::ScrollingTreeScrollingNodeDelegateMac):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::updateFromStateNode):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::handleWheelEvent):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::scrollPositionIsNotRubberbandingEdge const):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::allowsHorizontalStretching const):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::allowsVerticalStretching const):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::stretchAmount const):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::isPinnedOnSide const):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::shouldRubberBandOnSide const):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::updateScrollbarPainters):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::viewWillStartLiveResize):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::viewWillEndLiveResize):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::contentsSizeChanged):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::sizeDidChange):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.cpp:
(WebKit::RemoteScrollingCoordinatorProxy::viewWillStartLiveResize):
(WebKit::RemoteScrollingCoordinatorProxy::viewWillEndLiveResize):
(WebKit::RemoteScrollingCoordinatorProxy::contentsSizeChanged):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/ScrollingTreeFrameScrollingNodeRemoteMac.cpp:
(WebKit::ScrollingTreeFrameScrollingNodeRemoteMac::viewWillStartLiveResize):
(WebKit::ScrollingTreeFrameScrollingNodeRemoteMac::viewWillEndLiveResize):
(WebKit::ScrollingTreeFrameScrollingNodeRemoteMac::contentsSizeChanged):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/ScrollingTreeFrameScrollingNodeRemoteMac.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::viewWillStartLiveResize):
(WebKit::WebPageProxy::viewWillEndLiveResize):
(WebKit::WebPageProxy::contentsSizeChanged):
* Source/WebKit/UIProcess/WebPageProxy.h:

Canonical link: <a href="https://commits.webkit.org/261977@main">https://commits.webkit.org/261977@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/52fb1eacec97151a7b5d9a7a0768b4ea5a69f1b3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/29 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/31 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/30 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/26 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/25 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/28 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/29 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/28 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/19 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/28 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/27 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/20 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/29 "Built successfully") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/29 "3 api tests failed or timed out") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/19 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/23 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/20 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/26 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/26 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/26 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/21 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/20 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/26 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/page/scrolling/ScrollingTreeScrollingNodeDelegate.h:
```diff
@@ -50,6 +50,11 @@ class ScrollingTreeScrollingNodeDelegate {
+    virtual void viewWillEndLiveResize() { }
+    virtual void contentsSizeChanged() { }
+
+    
```
Extra line.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/mac/ScrollerPairMac.h:
```diff
@@ -76,8 +76,14 @@ class ScrollerPairMac {
+    void viewWillStartLiveResize();
+    void viewWillEndLiveResize();
+    void contentsSizeChanged();
+    bool inLiveResize() { return m_inLiveResize; }
```
const

nmoucht replied:
Fixed.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
```diff
@@ -92,10 +92,10 @@
 {
     if (!m_webPageProxy.hasRunningProcess())
         return;
+    m_webPageProxy.contentsSizeChanged();
```
`sizeDidChange()` is not the same as `contentsSizeChanged()`. The former means that the view was resized. The latter means that the scrolled contents got bigger or smaller.

nmoucht replied:
Changed these to viewSizeDidChange to make it more clear. contentsSizeChanged might need to be plumbed through the proxy scrollbar object from the web process.

smfr commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -2159,16 +2159,35 @@ void WebPageProxy::viewWillStartLiveResize()
     closeOverlayedViews();
+    
+#if ENABLE(ASYNC_SCROLLING) && PLATFORM(COCOA)
+    if (m_scrollingCoordinatorProxy)
```
I think it would be more natural to plumb these through the `DrawingAreaProxy` and have the `RemoteLayerTreeDrawingAreaProxy` overrides call into the scrolling coordinator.

nmoucht replied:
Fixed.

nmoucht committed:
[UI-side compositing] Resizing window doesn't cause scrollbars to show/change size
https://bugs.webkit.org/show_bug.cgi?id=253739
rdar://106385267

Reviewed by Simon Fraser.

Hook up ScrollerMac to UI-process resize events so that scrollbars
dynamically resize when resizing window. We should also consider subclassing
ScrollerMac under ScrollbarsControllerMac so we dont duplicate this
NScrollerImp interaction code.

* Source/WebCore/page/scrolling/ScrollingTree.cpp:
(WebCore::ScrollingTree::viewWillStartLiveResize):
(WebCore::ScrollingTree::viewWillEndLiveResize):
(WebCore::ScrollingTree::contentsSizeChanged):
* Source/WebCore/page/scrolling/ScrollingTree.h:
* Source/WebCore/page/scrolling/ScrollingTreeFrameScrollingNode.h:
* Source/WebCore/page/scrolling/ScrollingTreeScrollingNodeDelegate.h:
(WebCore::ScrollingTreeScrollingNodeDelegate::viewWillStartLiveResize):
(WebCore::ScrollingTreeScrollingNodeDelegate::viewWillEndLiveResize):
(WebCore::ScrollingTreeScrollingNodeDelegate::contentsSizeChanged):
* Source/WebCore/page/scrolling/mac/ScrollerPairMac.h:
* Source/WebCore/page/scrolling/mac/ScrollerPairMac.mm:
(WebCore::ScrollerPairMac::viewWillStartLiveResize):
(WebCore::ScrollerPairMac::viewWillEndLiveResize):
(WebCore::ScrollerPairMac::contentsSizeChanged):
* Source/WebCore/page/scrolling/mac/ScrollingTreeScrollingNodeDelegateMac.h:
* Source/WebCore/page/scrolling/mac/ScrollingTreeScrollingNodeDelegateMac.mm:
(WebCore::ScrollingTreeScrollingNodeDelegateMac::ScrollingTreeScrollingNodeDelegateMac):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::updateFromStateNode):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::handleWheelEvent):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::scrollPositionIsNotRubberbandingEdge const):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::allowsHorizontalStretching const):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::allowsVerticalStretching const):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::stretchAmount const):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::isPinnedOnSide const):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::shouldRubberBandOnSide const):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::updateScrollbarPainters):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::viewWillStartLiveResize):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::viewWillEndLiveResize):
(WebCore::ScrollingTreeScrollingNodeDelegateMac::contentsSizeChanged):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::sizeDidChange):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.cpp:
(WebKit::RemoteScrollingCoordinatorProxy::viewWillStartLiveResize):
(WebKit::RemoteScrollingCoordinatorProxy::viewWillEndLiveResize):
(WebKit::RemoteScrollingCoordinatorProxy::contentsSizeChanged):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/ScrollingTreeFrameScrollingNodeRemoteMac.cpp:
(WebKit::ScrollingTreeFrameScrollingNodeRemoteMac::viewWillStartLiveResize):
(WebKit::ScrollingTreeFrameScrollingNodeRemoteMac::viewWillEndLiveResize):
(WebKit::ScrollingTreeFrameScrollingNodeRemoteMac::contentsSizeChanged):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/ScrollingTreeFrameScrollingNodeRemoteMac.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::viewWillStartLiveResize):
(WebKit::WebPageProxy::viewWillEndLiveResize):
(WebKit::WebPageProxy::contentsSizeChanged):
* Source/WebKit/UIProcess/WebPageProxy.h:

Canonical link: https://commits.webkit.org/261977@main
